### PR TITLE
ISSUE-107: fix out of memory when fetching Top objects with lots of children

### PR DESCRIPTION
**Pull request by DiegoPino**
# What?

See #107. 
- Adds adaptive Range/offset based on what was the last number of Top objects v/s Children Rations
- Copies CMODEL mapping between Children/Top objects during setup
- Validates UUID Mapping v/s Node to Node Mapping (to avoid those being mixed up)
- Multiple cleanups, fixes Automatic UUID for Plugins that run via Batch (e.g Solr)

@karomabiles and @alliomeria even if this is safer, I feel it just became a bit slower so will need you both to check with me during our call the same collection being harvested via your current systems v/s with this patch. I can do a pre-test just to be sure of course and make you loose less time

### CODE REVIEW

File: src/Form/AmiMultiStepIngest.php [hunk 1]
```diff
@@ -150,12 +150,12 @@ public function buildForm(array $form, FormStateInterface $form_state) {
...
-        '#title' => $this->t('Webform'),
-        '#options' => $webform,
-        '#description' => $this->t('Columns are casted to ADO metadata (JSON) by passing/validating Data through an existing Webform'),
-      ];*/
+      /* $element_conditional['webform'] =[
+         '#type' => 'select',
+         '#title' => $this->t('Webform'),
+         '#options' => $webform,
+         '#description' => $this->t('Columns are casted to ADO metadata (JSON) by passing/validating Data through an existing Webform'),
+       ];*/
```
Comment (aksm) [inline src/Form/AmiMultiStepIngest.php:158]:
```suggestion
      /**
       * $element_conditional['webform'] = [
       *   '#type' => 'select',
       *   '#title' => $this->t('Webform'),
       *   '#options' => $webform,
       *   '#description' => $this->t('Columns are casted to ADO metadata (JSON) by passing/validating Data through an existing Webform'),
       * ];
       */
```

File: src/Form/AmiMultiStepIngest.php [hunk 2]
```diff
@@ -533,12 +576,18 @@ public function submitForm(array &$form, FormStateInterface $form_state) {
           $amisetdata->total_rows = $data['totalrows'];
         }
 
-        // We should probably add the UUIDs here right now.
-        $uuid_key = isset($amisetdata->adomapping['uuid']['uuid']) && !empty($amisetdata->adomapping['uuid']['uuid']) ? $amisetdata->adomapping['uuid']['uuid'] : 'node_uuid';
+        /* remove from the mappings any column used as file source */
```
Comment (aksm) [inline src/Form/AmiMultiStepIngest.php:579]:
```suggestion
        /** remove from the mappings any column used as file source */
```

File: src/Form/AmiMultiStepIngest.php [hunk 3]
```diff
@@ -533,12 +576,18 @@ public function submitForm(array &$form, FormStateInterface $form_state) {
...
+        else {
+          $uuid_key = isset($amisetdata->adomapping['uuid']['uuid']) && !empty($amisetdata->adomapping['uuid']['uuid']) ? $amisetdata->adomapping['uuid']['uuid'] : 'node_uuid';
+        }
         // We want to reset this value now
         $amisetdata->adomapping['uuid']['uuid'] = $uuid_key;
         if (!$plugin_instance->getPluginDefinition()['batch']) {
-          $fileid = $this->AmiUtilityService->csv_save($data, $uuid_key);
+          // We now pass also if auto UUID was chosen or not.
+          $fileid = $this->AmiUtilityService->csv_save($data, $uuid_key, $amisetdata->adomapping['autouuid'] ?? FALSE);
         } else {
```
Comment (aksm) [inline src/Form/AmiMultiStepIngest.php:591]:
```suggestion
        }
        else {
```

File: src/Plugin/ImporterAdapter/SolrImporter.php [hunk 4]
```diff
@@ -122,14 +122,23 @@ public function interactiveForm(array $parents, FormStateInterface $form_state):
...
+    $rows = $form_state->getValue(array_merge($parents,
+      ['solarium_config', 'rows']), 0);
+    if ($rows_form_state !== NULL) {
+      $rows = $rows_form_state;
+    }
     if (empty($rows)) {
       $form_state->setValue(['pluginconfig','ready'], FALSE);
     }
 
+    /* Calculate Max timeout based on PHP Settings */
```
Comment (aksm) [inline src/Plugin/ImporterAdapter/SolrImporter.php:135]:
```suggestion
    /** Calculate Max timeout based on PHP Settings */
```

File: src/Plugin/ImporterAdapter/SolrImporter.php [hunk 5]
```diff
@@ -231,12 +245,14 @@ public function interactiveForm(array $parents, FormStateInterface $form_state):
...
-        '#description' => $this->t('Total number of Rows to Fetch. Settings this to empty or null will prefill with the real Number Rows found by the Solr Query invoked. If you set this number higher than the actual results we will only fetch what can be fetched'),
+        '#title' => $this->t('Number of Rows for Top Objects'),
+        '#description' => $this->t('Total number of Rows to Fetch for top Objects. Settings this to empty or null will prefill with the real Number Rows found by the Solr Query invoked. If you set this number higher than the actual results we will only fetch what can be fetched.</br><em>Note:</em>Children objects are not being counted for and will add to your total spreadsheet row count if found.'),
         '#default_value' => $rows,
       ],
     ];
-
+    /* @TODO let's tell the user why it is not ready? */
+    /* Also todo, move mappings already done between children/parents/children */
+    /* That way previous decisions on any of each sides will stick on the other */
```
Comment (aksm) [inline src/Plugin/ImporterAdapter/SolrImporter.php:255]:
```suggestion
    /** 
     * @TODO let's tell the user why it is not ready?
     * Also todo, move mappings already done between children/parents/children
     * That way previous decisions on any of each sides will stick on the other
     */
```

File: src/Plugin/ImporterAdapter/SolrImporter.php [hunk 6]
```diff
@@ -445,7 +470,15 @@ public function getInfo(array $config, FormStateInterface $form_state, $page = 0
...
+        $pids = array_map('trim', explode("\n", $input));
+        $pids = array_filter($pids);
+        foreach ($pids as &$pid) {
+          $pid = $helper->escapePhrase('info:fedora/' . $pid);
+        }
+        $escaped = implode(' OR ', $pids);
+        $escaped = '('. $escaped .')';
+
         $query->setQuery('*:*');
         $query->createFilterQuery('collection_members')->setQuery('RELS_EXT_isMemberOfCollection_uri_s:'.$escaped .' OR RELS_EXT_isMemberOf_uri_s:'.$escaped );
```
Comment (aksm) [inline src/Plugin/ImporterAdapter/SolrImporter.php:483]:
```suggestion
        $query->createFilterQuery('collection_members')->setQuery('RELS_EXT_isMemberOfCollection_uri_s:' . $escaped . ' OR RELS_EXT_isMemberOf_uri_s:' . $escaped);
```

File: src/Plugin/ImporterAdapter/SolrImporter.php [hunk 7]
```diff
@@ -607,7 +640,7 @@ public function getInfo(array $config, FormStateInterface $form_state, $page = 0
         $form_state->setValue(['pluginconfig', 'ready'], FALSE);
       }
       if (((count(array_intersect_key($form_state->getValue(['pluginconfig', 'solarium_mapping','cmodel_children'], []), $cmodel_children)) == count($cmodel_children))
-        && count($cmodel_children) >= 1) || $form_state->getValue(['pluginconfig', 'solarium_mapping', 'collapse']) == TRUE) {
+          && count($cmodel_children) >= 1) || $form_state->getValue(['pluginconfig', 'solarium_mapping', 'collapse']) == TRUE) {
         $form_state->setValue(['pluginconfig', 'ready'], TRUE);
       } else {
```
Comment (aksm) [inline src/Plugin/ImporterAdapter/SolrImporter.php:645]:
```suggestion
      }
      else {
```

File: src/Plugin/ImporterAdapter/SolrImporter.php [hunk 8]
```diff
@@ -682,22 +740,39 @@ public function getData(array $config, $page = 0, $per_page = 20): array {
...
+        $pids = array_map('trim', explode("\n", $input));
+        $pids = array_filter($pids);
+        foreach ($pids as &$pid) {
+          $pid = $helper->escapePhrase('info:fedora/' . $pid);
+        }
+        $escaped = implode(' OR ', $pids);
+        $escaped = '('. $escaped .')';
         $query->setQuery('*:*');
+
         $query->createFilterQuery('collection_members')->setQuery('RELS_EXT_isMemberOfCollection_uri_s:'.$escaped .' OR RELS_EXT_isMemberOf_uri_s:'.$escaped );
```
Comment (aksm) [inline src/Plugin/ImporterAdapter/SolrImporter.php:762]:
```suggestion
        $query->createFilterQuery('collection_members')->setQuery('RELS_EXT_isMemberOfCollection_uri_s:' . $escaped . ' OR RELS_EXT_isMemberOf_uri_s:' . $escaped);
```

File: src/Plugin/ImporterAdapter/SolrImporter.php [hunk 9]
```diff
@@ -778,16 +853,18 @@ public function getData(array $config, $page = 0, $per_page = 20): array {
           if (in_array($sp_data[$resultset_iterator->key()]['rels_ext_hasmodel_uri'], static::MULTICHILDREN_CMODELS)) {
             $pids_to_fetch[$resultset_iterator->key()] = $sp_data[$resultset_iterator->key()]['pid'];
           }
+          // If deep traverse is enabled we should also mark collections here?
+          // Call getData recursevely?
+
         } catch (\Exception $exception) {
```
Comment (aksm) [inline src/Plugin/ImporterAdapter/SolrImporter.php:859]:
```suggestion
        }
        catch (\Exception $exception) {
```

File: src/Plugin/ImporterAdapter/SolrImporter.php [hunk 10]
```diff
@@ -903,8 +994,6 @@ protected function getDataChildren(array $config, SolariumClient $client, string
     ]);
 
     $resultset = $client->select($query);
-    // display the total number of documents found by solr
-    error_log('NumFound: ' . $resultset->getNumFound());
     if ($resultset->getNumFound() == 0) { return []; }
```
Comment (aksm) [inline src/Plugin/ImporterAdapter/SolrImporter.php:997]:
```suggestion
    if ($resultset->getNumFound() == 0) {
      return [];
    }
```

File: src/Plugin/ImporterAdapter/SolrImporter.php [hunk 11]
```diff
@@ -1106,17 +1238,31 @@ public static function fetchBatch(array $config, ImporterPluginAdapterInterface
       // And not the mess each doc returns
       $config['headers'] = !empty($amisetdata->column_keys) ? $amisetdata->column_keys : (!empty($config['headers']) ? $config['headers'] : []);
       $config['headerswithdata'] = $context['results']['processed']['headerswithdata'] ?? [];
-      $data = $plugin_instance->getData($config, $context['sandbox']['progress'] + $offset,
+
+      $data = $plugin_instance->getData($config, $nextoffset,
         $next_increment);
+
+      if (isset($data['nextforcedoffset']) && $data['nextforcedoffset'] !== NULL ) {
```
Comment (aksm) [inline src/Plugin/ImporterAdapter/SolrImporter.php:1245]:
```suggestion
      if (isset($data['nextforcedoffset']) && $data['nextforcedoffset'] !== NULL) {
```

File: src/Form/AmiMultiStepIngest.php (hunk 2 above)
Comment (DiegoPino) [reply src/Form/AmiMultiStepIngest.php:581]: I will leave this one as it is. Single commends can have /* */ or even //

**Comment by DiegoPino** [comment]
@aksm thanks for your review

Any reason why we need to use

` /**comments  */`

instead of 
`/* comments */`

Just checking my Drupal Coding standards knowledge. Was not aware that is a requirement? Please let me know

**Comment by aksm** [comment]
> @aksm thanks for your review
> 
> Any reason why we need to use
> 
> ` /**comments */`
> 
> instead of `/* comments */`
> 
> Just checking my Drupal Coding standards knowledge. Was not aware that is a requirement? Please let me know

@DiegoPino Sorry, I misunderstood the docs. I thought all of them had to start that way, but I see now that it's just docblocks.

**Comment by DiegoPino** [comment]
@aksm no worries. DCS are a lot to digest. Please check if I addressed all your concerns. If so we can move forward and approve and keep walking into other issues. Before merging I need to compare performance with @karomabiles and @alliomeria. I felt a small harvest last night went a bit slow. I can't see why/where it could have been slowed down so it might be just my internet, but still better to compare. Thanks a lot for reviewing!

**Comment by aksm** [comment]
@DiegoPino Looks good to me. Was only scanning for (my limited understanding of) DCS and anything obvious that might jump out at me. Thanks!

**Comment by DiegoPino** [comment]
@aksm if it looks good you need to approve it!

**Review by aksm — APPROVED** [review]

**Comment by DiegoPino** [comment]
There might be some Drupal 10.0 deprecations in some parts of AMI. Will deal with those in another pull. In the meantime merging this and we go from there. Thanks

**Pull request merged**